### ISSUE TITLE
Use a debounce function instead of using a refresh interval

### ISSUE THREAD

**Issue by Mogztter** [asciidoctor/asciidoctor-vscode#588]
Currently, we wait 2 seconds (default) even if this is the first event. We also do not cancel/ignore `onDidChangeTextDocument` events even if the user is still typing/updating the document. Instead, we should use a debounce function to delay the update until after 500 milliseconds (or so) have elapsed since the last time we receive an `onDidChangeTextDocument` event.

- https://redd.one/blog/debounce-vs-throttle
- https://lodash.com/docs/4.17.15

We should also introduce a new setting to update the preview only when the file is saved.

**Comment by ahus1** [comment]
Hi @Mogztter - thanks for starting this discussion. Reading the article you linked above, the current behavior is "debounce" in VS code if I understand it correctly, as the first update happens only after two seconds. I agree that the first event should be rendered immediately (1st key pressed -> immediate rendering; second delay after n milliseconds only), so the new behavior would be "throttle".

I've heard that users think something is broken when an update takes longer than one second. Therefore I'd vote for 500 ms (as developers are usually less patient than regular regular users).

**Comment by Mogztter** [comment]
> Reading the article you linked above, the current behavior is "debounce" in VS code if I understand it correctly, as the first update happens only after two seconds. I agree that the first event should be rendered immediately (1st key pressed -> immediate rendering; second delay after n milliseconds only), so the new behavior would be "throttle".

Yes and no, since throttle will ignore trailing events. For instance, if I type "a" and "b" within 500ms or less, it will trigger an event on the first letter "a" and ignore "b".
I think we should not trigger an event while the user is still typing.

I believe it's possible to trigger an event on the first key pressed using the leading option (or the `flush` function): https://lodash.com/docs/4.17.15#debounce

**Comment by ahus1** [comment]
>  I think we should not trigger an event while the user is still typing.

So you want the user to stop typing before refreshing the view. I've usually advocated to keep the preview as much in sync with the editor as possible, which includes refreshing the preview quite often. When would a deferred update be better?

**Comment by Mogztter** [comment]
> When would a deferred update be better?

It mostly about performance. If refreshing the preview (i.e., load and convert the AsciiDoc document and refresh the webview) takes little to no resource then, we should indeed refresh often.

Since we cannot cancel the "refresh process" (it might be possible using WebWorker?) it's a waste to load and convert an outdated AsciiDoc document because the user is still "actively" typing.

Another idea would be:

- first letter -> no delay
- actively typing -> at most every 500ms
- stop typing (i.e., last/trailing event) -> 500ms delay

**Comment by ahus1** [comment]
+1 for this idea. 

Another constraint: If the rendering of a preview is currently in progress (and it can't be cancelled as you mentioned), no concurrent rendering should start. So we might be at something like "next rendering will start 500ms after the previous rendering is complete when there was a change to the docs in the meantime"
